[Fix for qualification tests] - Bluetooth: controller:  Add filtering of periodic advertising sync. by CTE type

ppryga-nordic opened the pull request:
Periodic advertisement synchronization may be filtered by included
CTE type. If particula CTE type is not allowed then deepening on
filtering policy:
- if filtering policy is off synchronization if terminated
- if filtering policy is on synchronization is continued to
  synchronize with other device from allowed advertisements list.
If synchronization is established and peer device changes CTE type
to one that is not allowed, synchronization should be maintained.

Changes in LLL:
There are two new execution paths. First one is executed when
synchronization is created. In this case CTEInline is enabled
to parse PDU for CTEInfo field. In this execution path CTE
type is verified. Second execution path does not include
parsing PDU for CTEInfo and verification of CTE type.

Information about sync allowed is added to node_rx instance
that transports received PDU data. In case the sync has to be
terminated the node_rx will not hold PDU data.

Also done event is extended with information about sync
termination if CTE type is not allowed and filtering
policy is off.

Changes in ULL:
The NODE_RX_TYPE_SYNC is used to transport information that:
- Sync is established. In such situation the node_rx
  includes data related with received PDU
- Sync scanning is terminated.
In first case ULL will generate NODE_RX_TYPE_SYNC_REPORT
after sending NODE_RX_TYPE_SYNC.

Also EVENT_DONE_EXTRA_TYPE_SYNC handling has additional
execution path that terminates sync scanning if requested
by lower link layer. In other case it adjusts sync scan
window and maintains timeout as usual.

ppryga-nordic committed:
Bluetooth: hci: Add helper macros for ver of disallowed CTE types

The commit adds helper macros for verification of disallowed
CTE types when periodic advertising synchornization is created.

The macros are added here, because they are directly related
with values specified by BT Core 5.1 specification.

Signed-off-by: Piotr Pryga <piotr.pryga@nordicsemi.no>

ppryga-nordic committed:
Bluetooth: hci: Add handling of allowed CTE types for per sync create

Add missing code responsible for handling of allowed CTE types
in HCI_LE_Periodic_Advertising_Create_Sync command.

Signed-off-by: Piotr Pryga <piotr.pryga@nordicsemi.no>

ppryga-nordic committed:
Bluetooth: radio: move radio_df_cte_inline_set_enabled to radio.c

To enable runtime parsing of PDU to find CTEInfo field CTEINLINE mode
has to be enabled. Thanks to that it is possible to verify if the PDU
has allowed CTE type e.g. for periodic advertising synchornization.
To run CTEInfo parsing other parametrers of CTEINLINE are not relevant.
If Radio is set to disable after PDU END event the CTE sampling
will not be processed.

The commit moves the radio_df_cte_inline_set_enable function to make
it accessible even the direction finding features are disabled.

Signed-off-by: Piotr Pryga <piotr.pryga@nordicsemi.no>

ppryga-nordic committed:
Bluetooth: controller: rework per scan lll to allow filt by CTE type

Periodic advertisement synchronization may be filtered by CTE type.
If particular CTE type is not allowed then depening on filtering policy:
- if filtering policy is off synchronization if terminated
- if filtering policy is on synchronization is continued to
  synchonize with another device from allowed adverisements list.
If synchronization is established and peer device changes CTE type
to one that is not allowed, synchronization should be maintained.

There are two new execution paths. First one is executed when
synchronization is created. In this case CTEILINE is enabled
to parse PDU for CTEInfo field. In this execution path CTE
type is verified. Second execution path does not include
parsing PDU for CTEInfo and verification of CTE type.

Information about sync allowed is added to node_rx instance
that transports received PDU data. In case the sync has to be
terminated the node_rx will not hold PDU data.

Also done event is extended with information about sync
termination if CTE type is not allowed and filtering
policy is off.

Signed-off-by: Piotr Pryga <piotr.pryga@nordicsemi.no>

ppryga-nordic committed:
Bluetooth: controller: ULL: add per sync filtering by CTE type

Follow up on changes in lower link layer to add filtering
of periodic advertisements synchronization by CTE type.

The NODE_RX_TYPE_SYNC is used to transport information that:
- Sync is established. In such situation the node_rx
  includes data related with received PDU
- Sync scanning is terminated.
In first case ULL will generate NODE_RX_TYPE_SYNC_REPORT
after sending NODE_RX_TYPE_SYNC.

Also EVENT_DONE_EXTRA_TYPE_SYNC handling has additional
execution path that terminates sync scanning if requested
by lower link layer. In other case it adjusts sync scan
window and maintains timeout as usual.

Signed-off-by: Piotr Pryga <piotr.pryga@nordicsemi.no>

cvinayak commented on subsys/bluetooth/controller/ll_sw/lll.h:
```diff
@@ -399,7 +402,11 @@ struct event_done_extra {
 		struct {
 			uint16_t trx_cnt;
-			uint8_t  crc_valid;
+			uint8_t  crc_valid:1;
+			/* Used to inform ULL that periodic advertising sync scan should be
+			 * terminated.
+			 */
+			uint8_t  sync_term:1;
```
```suggestion
			uint8_t  crc_valid:1;
			/* Used to inform ULL that periodic advertising sync scan should be
			 * terminated.
			 */
#if defined(CONFIG_BT_CTLR_SYNC_PERIODIC)
			uint8_t  sync_term:1;
#endif /* CONFIG_BT_CTLR_SYNC_PERIODIC */
```

ppryga replied:
done

cvinayak commented on include/bluetooth/hci.h:
```diff
@@ -1430,11 +1430,16 @@ struct bt_hci_cp_le_ext_create_conn {
 #define BT_HCI_LE_PER_ADV_CREATE_SYNC_FP_USE_LIST               BIT(0)
 #define BT_HCI_LE_PER_ADV_CREATE_SYNC_FP_REPORTS_DISABLED       BIT(1)
 
+#define BT_HCI_LE_PER_ADV_CREATE_SYNC_CTE_TYPE_NO_FILTERING 0
```
indent the `0`

ppryga replied:
Done

cvinayak commented on subsys/bluetooth/controller/ll_sw/lll_sync.h:
```diff
@@ -34,6 +34,7 @@ struct lll_sync {
 
 	uint8_t phy:3;
 	uint8_t is_rx_enabled:1;
+	uint8_t cte_type:4;
```
Conditional compile?

ppryga replied:
Why conditionally compile? It is a feature of per. adv. sync not a feature of direction finding.

ppryga replied:
Then:
1. Whole feature should configurable to enable/disable.
2. The code may be left as is, because the BT Core Spec 5.0 reserves the Sync_CTE_Type (as `Unused`) parameter of HCI_LE_Periodic_Advertising_Create_Sync command. It is imposed on host to always send value zero in this parameter  to comply with the 5.0 Core spec. So it would be handled as no filtering by this implementation.
Besides bigger code and more instructions executed behavior should be the same.

cvinayak commented on subsys/bluetooth/controller/ll_sw/ull.c:
```diff
@@ -1827,6 +1827,18 @@ void ull_prepare_dequeue(uint8_t caller_id)
 	}
 }
 
+struct event_done_extra *ull_event_done_extra_dequeue(void)
+{
+	struct node_rx_event_done *evdone;
+
+	evdone = MFIFO_DEQUEUE_PEEK(done);
+	if (!evdone) {
+		return NULL;
+	}
+
+	return &evdone->extra;
+}
+
```
Isn't this duplicate of `ull_event_done_extra_get()` ?
```suggestion
```

ppryga replied:
Ah.. sorry it was left over from previous iteration of the change

cvinayak commented on subsys/bluetooth/controller/ll_sw/ull_sync_types.h:
```diff
@@ -12,6 +12,9 @@
 struct ll_sync_iso;
 #endif /* CONFIG_BT_CTLR_SYNC_ISO */
 
+/* Periodic advertisements synchronization status. */
+enum sync_status { SYNC_STAT_FOUND, SYNC_STAT_READY, SYNC_STAT_TERM };
+
```
```suggestion
/* Periodic advertisements synchronization status. */
enum sync_status {
	SYNC_STAT_FOUND,
	SYNC_STAT_READY,
	SYNC_STAT_TERM
};

```

ppryga replied:
enum moved to `lll_sync.h`

cvinayak commented on subsys/bluetooth/controller/ll_sw/ull.c:
```diff
@@ -2407,6 +2416,14 @@ static inline int rx_demux_rx(memq_link_t *link, struct node_rx_hdr *rx)
 		ull_scan_aux_release(link, rx);
 	}
 	break;
+#if defined(CONFIG_BT_CTLR_SYNC_PERIODIC)
+	case NODE_RX_TYPE_SYNC_REPORT:
+	{
+		(void)memq_dequeue(memq_ull_rx.tail, &memq_ull_rx.head, NULL);
+		ull_sync_report(link, rx);
+	}
+	break;
+#if defined(CONFIG_BT_CTLR_DF_SCAN_CTE_RX)
```
`ull_sync_report`  will check `ftr->sync_status` for every `NODE_RX_TYPE_SYNC_REPORT`, this beats the purpose of adding `lll_sync_create_prepare`. Instead, why not prepare and send `NODE_RX_TYPE_SYNC` from  the `lll_sync_create_prepare` control path?

ppryga replied:
I've changed approach. Now The NODE_RX_TYPE_SYNC is used to transport information about sync established and first received PDU. That is handled by `ull_sync_established_report` (I've renamed the function) function. NODE_RX_TYPE_SYNC_REPORT are handled as they were in the past.

cvinayak commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll_sync.c:
```diff
@@ -36,24 +36,38 @@
 
 #include "ll_feat.h"
 
+#include "ull_internal.h"
+#include "ull_sync_types.h"
```
Should not use ULL in LLL, if there is any member that  needs access by LLL, should be placed in LLL types or extern interface be added to get it.

ppryga replied:
Moved `enum sync_status` to lll_sync.h and removed ULL includes from the file.

cvinayak commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll_sync.c:
```diff
@@ -277,10 +366,9 @@ static int prepare_cb(struct lll_prepare_param *p)
-	       ((EVENT_JITTER_US + EVENT_TICKER_RES_MARGIN_US +
-		 lll->window_widening_event_us) << 1) +
-	       lll->window_size_event_us;
+	hcto = ((EVENT_JITTER_US + EVENT_TICKER_RES_MARGIN_US + lll->window_widening_event_us)
+		<< 1) +
+	       remainder_us + lll->window_size_event_us;
```
Any reason to change the order?

Original order wass, the packet/event  timer takes `remainder_us` to apply the fine microsecond interval remainder, then window widening plus the jitters to be considered, finally the window size.

ppryga replied:
Nope, just wanted to avoid constant updates of re-formatted code by clang-format. As it may be seen above, rearrangement didn't help either. 
Reverted to original form.

ppryga-nordic committed:
samples: Bluetooth: df: enable filtering of per adv by CTE type

Enable filtering of periodic advertisements to synchronize
with advertisenemts that include CTE.

Signed-off-by: Piotr Pryga <piotr.pryga@nordicsemi.no>

ppryga commented:
Addressed comments from @cvinayak 
Fixed issues with not building tests.
Added commit that enables filtering in DF RX sample.

I've realized one thing. The implementation requires existence of DFE in Radio peripheral. 
That means filtering will not work with other nRF boards.
That may be addressed by traversing through received PDU to find CTEInfo there.
Though, it may be not good to implement it in LLL context.
At the end of the day, the software based filtering may require additional implementation in ULL.

One more thing, need to test the functionality behavior in situation a sync is lost and recovered.

ppryga commented:
Fixed compliance issue.

ppryga-nordic committed:
Bluetooth: controller: Add per sync filt by CTE type for SOC w/o DFE

First implementation of periodic advertising sync filtering
requires existence of Direction Finding Extension in Radio
peripheral.
To add the filtering support for other Nodric SOCs software
based PDU traversing for CTEInfo should be implemented.

In case there is no DFE in Radio peripheral, actual filtering
is done in ULL.

The commit provides necessary changes to previous solution.

Signed-off-by: Piotr Pryga <piotr.pryga@nordicsemi.no>

ppryga commented:
Added new commit that delivers support for per. adv. sync filtering for Nordic SOCs that do not have Direction Finding Extension in Radio peripheral.

The commit changes code introduces in commits that were already in the PR.
I've done it that way to show actual changes done to provide the feature where is no HW support for parsing PDU for CTEInfo.

ppryga commented:
Rebased onto main due to merge of https://github.com/zephyrproject-rtos/zephyr/pull/38184

Thalley commented on include/bluetooth/hci.h:
```diff
@@ -1430,11 +1430,16 @@ struct bt_hci_cp_le_ext_create_conn {
 #define BT_HCI_LE_PER_ADV_CREATE_SYNC_CTE_TYPE_ONLY_CTE         BIT(4)
+/* Constants to check correctness of CTE type */
+#define BT_HCI_LE_PER_ADV_CREATE_SYNC_CTE_TYPE_ALLOWED_BITS 4
+#define BT_HCI_LE_PER_ADV_CREATE_SYNC_CTE_TYPE_INVALID_VALUE                                       \
```
This seems more typical for this file, but it's okay as is as well 

```suggestion
#define BT_HCI_LE_PER_ADV_CREATE_SYNC_CTE_TYPE_INVALID_VALUE \
```

ppryga replied:
Corrected

Thalley commented on subsys/bluetooth/controller/Kconfig.df:
```diff
@@ -26,6 +26,8 @@ config BT_CTLR_DF_ANT_SWITCH_2US_SUPPORT
 config BT_CTLR_DF_ANT_SWITCH_1US_SUPPORT
 	bool
 
+config BT_CTLR_CTEINLINE_SUPPORT
+	bool
```
```suggestion
config BT_CTLR_CTEINLINE_SUPPORT
	bool

```

ppryga replied:
Added empty line

Thalley commented on subsys/bluetooth/controller/hci/hci.c:
```diff
@@ -3285,10 +3285,13 @@ static void le_per_adv_create_sync(struct net_buf *buf, struct net_buf **evt)
-				cmd->addr.a.val, skip, sync_timeout,
-				cmd->cte_type);
-
+	if ((cmd->cte_type & BT_HCI_LE_PER_ADV_CREATE_SYNC_CTE_TYPE_INVALID_VALUE) ==
+	    BT_HCI_LE_PER_ADV_CREATE_SYNC_CTE_TYPE_INVALID_VALUE) {
```
Since `BT_HCI_LE_PER_ADV_CREATE_SYNC_CTE_TYPE_INVALID_VALUE` is a bitmask, shouldn't this have been 

```suggestion
	if ((cmd->cte_type & BT_HCI_LE_PER_ADV_CREATE_SYNC_CTE_TYPE_INVALID_VALUE) != 0) {
```

Instead?

ppryga replied:
Yes it should, good point!

Thalley commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll_sync.c:
```diff
@@ -158,25 +183,66 @@ void lll_sync_aux_prepare_cb(struct lll_sync *lll,
+	 * Zero value in CTETime field of CTESTATUS may be used to distinguish between PDU that
+	 * includes CTEInfo or not. Allowed range for CTETime is 2-20.
+	 */
+	if (rx_cte_time == 0 && cte_type_mask & BT_HCI_LE_PER_ADV_CREATE_SYNC_CTE_TYPE_ONLY_CTE) {
```
```suggestion
	if (rx_cte_time == 0 && (cte_type_mask & BT_HCI_LE_PER_ADV_CREATE_SYNC_CTE_TYPE_ONLY_CTE) != 0) {
```

ppryga replied:
Added if not equal to zero

Thalley commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll_sync.c:
```diff
@@ -158,25 +183,66 @@ void lll_sync_aux_prepare_cb(struct lll_sync *lll,
+	}
+
+	if (rx_cte_time > 0) {
+		if (cte_type_mask & BT_HCI_LE_PER_ADV_CREATE_SYNC_CTE_TYPE_NO_CTE) {
```
```suggestion
		if ((cte_type_mask & BT_HCI_LE_PER_ADV_CREATE_SYNC_CTE_TYPE_NO_CTE) != 0) {
```

Thalley replied:
This comment applies multiple places if you want to follow the MISRA guidelines about strict typing. (14.4. in https://docs.zephyrproject.org/latest/contribute/coding_guidelines/index.html#main-rules)

ppryga replied:
Updated lines where cte_type_mask if used in `if` statements.

Thalley commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll_sync.c:
```diff
@@ -188,12 +254,114 @@ static int prepare_cb(struct lll_prepare_param *p)
+	lll = p->param;
+
+	/* Calculate the current event latency */
+	lll->skip_event = lll->skip_prepare + p->lazy;
+
+	/* Calculate the current event counter value */
+	event_counter = lll->event_counter + lll->skip_event;
```
I realize that this function has just been moved and this is outside the scope of this PR, but is there any chance that these additions could cause an overflow? We are adding two `uint16_t`s.

Thalley replied:
Could you add a comment regarding that here?

ppryga replied:
Yyy... comment regarding what? There is a response above, did I missed something?

Thalley replied:
I meant adding a comment in the code, stating that the event_counter may overflow, but that is it okay, as it can wrap.

Thalley commented on subsys/bluetooth/controller/ll_sw/ull.c:
```diff
@@ -2398,15 +2398,23 @@ static inline int rx_demux_rx(memq_link_t *link, struct node_rx_hdr *rx)
 		ull_scan_aux_setup(link, rx);
-	}
-	break;
+	} break;
```
I guess this is your automated tool's work?

I am not sure if this follows the coding guideline of
> Note that the closing brace is empty on a line of its own, except in the cases where it is followed by a continuation of the same statement, ie a while in a do-statement or an else in an if-statement:

(from https://www.kernel.org/doc/html/v4.10/process/coding-style.html)

In either way, this line should probably not have been changing in this PR.

ppryga replied:
Yes it was clang-format change. I've moved it to new line.

Thalley commented on subsys/bluetooth/controller/ll_sw/ull_sync.c:
```diff
@@ -527,84 +595,85 @@ void ull_sync_done(struct node_rx_event_done *done)
+			sync->timeout_expire = 0U;
+		}
+		/* If anchor point not sync-ed, start timeout countdown, and break skip if any */
+		else {
+			if (!sync->timeout_expire) {
```
```suggestion
		else if (!sync->timeout_expire) {
```

ppryga replied:
Corrected

Thalley commented on subsys/bluetooth/controller/ll_sw/ull_sync.c:
```diff
@@ -846,3 +910,26 @@ static void ticker_update_op_status_give(uint32_t status, void *param)
+	com_hdr = &pdu->adv_ext_ind;
+	hdr = &com_hdr->ext_hdr;
+	/* Skip flags */
+	dptr = hdr->data;
+
+	if (!com_hdr->ext_hdr_len || !hdr->cte_info) {
+		return NULL;
+	}
```
```suggestion
	if (!com_hdr->ext_hdr_len || !hdr->cte_info) {
		return NULL;
	}
	
	dptr = hdr->data;
```

ppryga replied:
Corrected

Thalley commented on subsys/bluetooth/controller/ll_sw/ull_sync_types.h:
```diff
@@ -21,6 +21,13 @@ struct ll_sync_set {
 	uint16_t volatile timeout_reload; /* Non-zero when sync established */
 	uint16_t timeout_expire;
 
+#if !defined(CONFIG_BT_CTLR_CTEINLINE_SUPPORT)
```
The `sync_term` in subsys/bluetooth/controller/ll_sw/lll.h is guarded as 
```
#if defined(CONFIG_BT_CTLR_SYNC_PERIODIC) && defined(CONFIG_BT_CTLR_CTEINLINE_SUPPORT)
```

while this only guards with `!defined(CONFIG_BT_CTLR_CTEINLINE_SUPPORT)`, which seems to be inconsistent.

ppryga replied:
I didn't add `CONFIG_BT_CTLR_SYNC_PERIODIC` because `ll_sync_set` is used only if `CONFIG_BT_CTLR_SYNC_PERIODIC` is enabled.

Thalley commented on subsys/bluetooth/controller/ll_sw/ull_sync_types.h:
```diff
@@ -33,6 +40,8 @@ struct ll_sync_set {
 		};
 	} node_rx_lost;
 
+	struct node_rx_hdr *node_rx_sync_estab;
+
```
Unused variable?

ppryga replied:
It is used but in commits following this commit e.g. https://github.com/zephyrproject-rtos/zephyr/blob/3c3805142656758a24fdf853cbcc9f8fed3159ce/subsys/bluetooth/controller/ll_sw/ull_sync.c#L158

ppryga commented:
@Thalley @cvinayak all your comments are addressed.

Thalley commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll_sync.c:
```diff
@@ -531,37 +643,141 @@ static int isr_rx(struct lll_sync *lll, uint8_t node_type, uint8_t *crc_ok)
+	/* Handle regular PDU reception if CTE type is acceptable */
+	if (sync_ok == SYNC_STAT_ALLOWED) {
+		err = isr_rx(lll, NODE_RX_TYPE_SYNC, crc_ok, rssi_ready, SYNC_STAT_ALLOWED);
+		if (err == -EBUSY) {
+			return;
+		}
```
What if there's another error?

ppryga replied:
There is no possibility for other error code except `0` and `-EBUSY`. I've added a header before isr_rx function with information about possible return values.

FYI, zero is returned in case there is no chained PDU or chained PDU is spaced long enough to schedule its reception by ULL. `-EBUSY` means there is a chained PDU but too close for ULL scheduling, so it is done directly from LLL.

Thalley replied:
> There is no possibility for other error code except 0 and -EBUSY

Isn't that a dangerous thing to assume, in case someone else updates the function at some point? I'm thinking that it would be "safer" to add an ASSERT to ensure that any other return values are "handled" or at least considered.

ppryga replied:
Well, I'm big fan of defensive programming, nevertheless this is an ISR and I here every instruction counts. I'd like to have an opinion from @cvinayak about it.

Thalley replied:
Aren't ASSERT compiled out for production builds, and thus safe to add? 

In any ways, if the function can only return 0 and -EBUSY, I guess the compiler would be able to optimize it anyhow, right?

ppryga replied:
> Aren't ASSERT compiled out for production builds, and thus safe to add?

In theory the LL_ASSERTs should be removed in release builds. 

> In any ways, if the function can only return 0 and -EBUSY, I guess the compiler would be able to optimize it anyhow, right?

This is just ques what compiler may or may not do.

I have a vague memory that I've been asked to remove LL_ASSERT from code. Because of that I asked @cvinayak about his point of view about it.

Thalley replied:
> anyone adding new return value in the future will definitely change the code to handle it.
In my opinion that's a optimistic expectation, but I suppose we'll have some CI tests that will catch it. 

I won't be blocking the PR based on this, but I do think it's a optimistic way of handling it.

ppryga replied:
If @Thalley does not push to use assert here I'll go with @cvinayak.

Just to be clear, I'm fan of defensive programming. If I were main contributor to the code I'll go with @Thalley suggestion.

Thalley commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll_sync.c:
```diff
@@ -531,37 +643,141 @@ static int isr_rx(struct lll_sync *lll, uint8_t node_type, uint8_t *crc_ok)
+	} else if (sync_ok == SYNC_STAT_TERM) {
+		struct node_rx_pdu *node_rx;
+
+		node_rx = ull_pdu_rx_alloc_peek(3);
+		if (node_rx) {
+			struct node_rx_ftr *ftr;
+
+			ull_pdu_rx_alloc();
```
Can you elaborate on why we need `peek(3)` for a single alloc?

ppryga replied:
`ull_pdu_rx_alloc_peek` name is a bit unfortunate. It peeks one buffer but only in case there are `x` buffers available. This is to ensure there are buffers for other use cases. 
In this case the verification is to ensure there are nodes for: reporting received PDU (the one that is allocated), 1x to receive a PDU in a connection, 1x to receive empty PDU.

I'll add a comment above.

BTW. I'm not going to change the function name ;-)

Thalley commented on subsys/bluetooth/controller/ll_sw/ull.c:
```diff
@@ -2407,6 +2407,15 @@ static inline int rx_demux_rx(memq_link_t *link, struct node_rx_hdr *rx)
 	}
 	break;
+#if defined(CONFIG_BT_CTLR_SYNC_PERIODIC)
+	case NODE_RX_TYPE_SYNC:
+
+	{
+		(void)memq_dequeue(memq_ull_rx.tail, &memq_ull_rx.head, NULL);
+		ull_sync_established_report(link, rx);
+	}
+	break;
```
```suggestion
	case NODE_RX_TYPE_SYNC:
		(void)memq_dequeue(memq_ull_rx.tail, &memq_ull_rx.head, NULL);
		ull_sync_established_report(link, rx);
		break;
```

ppryga replied:
I've just removed empty line after `case NODE_RX_TYPE_SYNC` to follow convention in other case statements in the function.

Thalley replied:
That's fine; didn't see that the other cases used brackets as well.

Thalley commented on subsys/bluetooth/controller/ll_sw/ull_sync.c:
```diff
@@ -848,3 +914,27 @@ static void ticker_update_op_status_give(uint32_t status, void *param)
+
+	com_hdr = &pdu->adv_ext_ind;
+	hdr = &com_hdr->ext_hdr;
+	/* Skip flags */
```
Misplaced comment? Or is it stating that we are in fact skipping flags in this functions? If the latter, can you elaborate (in the comment itself), as to why.

ppryga replied:
Yes the comment should be in line 932. I've moved it and extended a bit.
Also I've added comment why there are two LL_ASSERS below (lines 935,936).

ppryga commented:
I've addressed comments from @Thalley.
I've fixed one issue that caused bsim_advx tests to fail. It was related with wrong access to memory in `pdu_cte_info_get` function.
There is another issue in bsim_advx. It is related with update of SyncInfo field in AUX_ADV_IND PDU by `mfy_sync_offset_get`. 
This function is called by mayfly after `lll_adv_aux_prepare`. The calls are chained in scheduled for run by `ull_adv_aux_start`. 
It may occur that `lll_adv_aux_prepare` is put into wait queue due to other prepare pending. The `mfy_sync_offset_get` is called afterwards.
During execution of `mfy_sync_offset_get` it gets access to PDU where SyncInfo is going to be updated. The function uses `lll_adv_aux_data_curr_get` to do that. If the function `mfy_sync_offset_get` is called after delay `lll_adv_aux_prepare` it may occur that former advertising PDU is updated with new SyncInfo data. Later on when `lll_adv_aux_prepare` is run, a PDU, accessed by `lll_adv_aux_data_latest_get`, is set to be transmitted it has wrong SyncInfo data.

To solve that problem new function should be implemented that would return latest PDU but will not update pointers in double buffer used to store PDU. The double buffer pointers may be updated only once in prepare calback.
Such function is implemented as part of this PR: https://github.com/zephyrproject-rtos/zephyr/pull/38295.

So to make this PR to pass the CI the PR https://github.com/zephyrproject-rtos/zephyr/pull/38295 must be merged before.

ppryga commented on subsys/bluetooth/controller/Kconfig.df:
```diff
@@ -26,6 +26,9 @@ config BT_CTLR_DF_ANT_SWITCH_2US_SUPPORT
 config BT_CTLR_DF_ANT_SWITCH_1US_SUPPORT
 	bool
 
+config BT_CTLR_CTEINLINE_SUPPORT
```
This is not a term taken from BT spec. It is a term taken from nRF Radio documentation. It it actually a register name that allows to configure runtime PDU parsing to search for CTEInfo field.
I had to add this Kconfig option to allow use of the functionality without enabling complete DF functionality.

The name was intentionally `CTEINLINE` without a space to be the same as is used in nRF Radio documentation.

Thalley reviewed (approved):
My comments have been addressed

ppryga commented on subsys/bluetooth/controller/ll_sw/nordic/hal/nrf5/radio/radio.c:
```diff
@@ -1508,3 +1508,38 @@ void radio_ar_resolve(uint8_t *addr)
 }
+
+/* @brief Function configures CTE inline register to start sampling of CTE
+ *        according to information parsed from CTEInfo filed of received PDU.
```
Corrected

ppryga commented on subsys/bluetooth/controller/ll_sw/nordic/hal/nrf5/radio/radio.c:
```diff
@@ -1508,3 +1508,38 @@ void radio_ar_resolve(uint8_t *addr)
+/* @brief Function configures CTE inline register to start sampling of CTE
+ *        according to information parsed from CTEInfo filed of received PDU.
+ *
+ * @param[in] cte_info_in_s1    Informs where to expect CTEInfo filed in PDU:
```
Corrected

ppryga commented:
@cvinayak could you take a look on the PR? I've resolved typos related comments. Also I have rebased it on latest main after merge: https://github.com/zephyrproject-rtos/zephyr/pull/38515.

ppryga commented:
Fixed warning in lll_sync.c file.

ppryga-nordic committed:
tests: Bluetooth: bsim_test_iso: Change wait time in BIG receiver

The receiver and transmitter in the test are synchronized by
use of sleep functions. The change in handling of periodic
advertising synchronized event caused missmach in waiting
times. Receiver is notified about established synchronization
later that it was in the past. Due to that the wait for end
of transmitter operation was too long.

Temporary fix for the problem is decrease of receiver sleep
time by one periodic advertising interval.

Signed-off-by: Piotr Pryga <piotr.pryga@nordicsemi.no>

ppryga commented:
Added commit to fix failing `bsim_test_iso`.

ppryga commented:
Rebased after merge https://github.com/zephyrproject-rtos/zephyr/pull/38338. 
Resolved conflicts.
Found error in hci.h macro: 
```
#define BT_HCI_LE_PER_ADV_CREATE_SYNC_CTE_TYPE_INVALID_VALUE \
	(~BIT_MASK(BT_HCI_LE_PER_ADV_CREATE_SYNC_CTE_TYPE_ALLOWED_BITS))
```
There was missing bit-wise negation.

ppryga commented on subsys/bluetooth/controller/ll_sw/lll_sync.h:
```diff
@@ -12,6 +15,8 @@ struct lll_sync {
 
 	uint8_t phy:3;
 	uint8_t is_rx_enabled:1;
+	uint8_t cte_type:4;
+	uint8_t filter_policy:1;
```
Regarding the members, in my humble opinion the right place to define them is in `lll_sync`. Even though storing them in `ll_scan_set.per_scan` would not introduce error or race conditions, the first location for such members in seems to be `lll_sync`. Besides that, or even more importantly, access to `ll_scan_set` may influence number of instructions in LLL execution context. 
Before merge of your PR related with per. adv. list the `filter_policy` was one of removed members from `per_scan` in this PR. 

Referring to `lll_sync_cte_is_allowed` in my understanding of BT Core spec 5.1 the cte_type filtering of periodic advertisements is not dependent on DF features. 
>4.4.3.4 Scanning for periodic advertisements.
>[...]
>The Host may instruct the Controller not to synchronize to periodic advertising
trains with certain types of Constant Tone Extensions or without a Constant
Tone Extension. If the Controller receives an AUX_SYNC_IND with such a
Constant Tone Extension while synchronizing, the synchronization attempt has
failed and shall cease. Once synchronized, the presence or type of Constant
Tone Extension shall not affect synchronization.

cvinayak replied:
All I am saying is, applications like ISO Synchronized Receiver application do not need CTE filtering and will use `Sync_CTE_Type` value set to `0`. (`CONFIG_BT_CTLR_FILTER_ACCESS_LIST`, `CONFIG_BT_CTLR_PRIVACY`, and `CONFIG_BT_CTLR_SYNC_PERIODIC_ADV_LIST` are Kconfig options available to applications today to exclude respective filtering).

What I understand from the specification of `7.8.67 LE Periodic Advertising Create Sync command`, the `Options` field and `Sync_CTE_Type` are independent. I.e. irrespective of whether Periodic Advertiser List is used, filtering based on CTE is solely determined by `Sync_CTE_Type`.

> The Sync_CTE_Type parameter specifies whether to only synchronize to
periodic advertising with certain types of Constant Tone Extension (a value of 0
indicates that the presence or absence of a Constant Tone Extension is irrelevant).

Hence, the changes requested is to add a Kconfig around the CTE type filtering implementation. And based on the new information I became aware of from the specification, remove use of implementation defined `filter_policy` field added here in these lines.

ppryga-nordic committed:
Bluetooth: controller: Make per adv filtering by CTE cond compilable

The filtering of periodic advertisements by scanner may be not needed
in certain situations e.g. while use of periodic advertising by BT ISO.
To make the code smaller and avoid execution of not needed code the
functionality will be conditionally compilable. It may be enabled
or disabled by use of CONFIG_BT_CTLR_SYNC_PERIODIC_CTE_TYPE_FILTERING
Kconfig option.

Signed-off-by: Piotr Pryga <piotr.pryga@nordicsemi.no>
